pkg/hub/controllers/finalizercontroller/finalize_controller.go

```diff
@@ -0,0 +1,278 @@
+		return nil
+	}
+	clusterName := strings.TrimSuffix(name, keySuffix)
+
+	// ignore role/rolebinding if namespace does not match with cluster name
+	if namespace != clusterName {
+		return nil
+	}
+
+	ns, err := m.namespaceLister.Get(namespace)
+	if err != nil {
+		return err
+	}
+
+	role, rolebinding, err := m.getRoleAndRoleBinding(namespace, name)
+	if err != nil {
+		return err
+	}
+
+	err = m.syncRoleAndRoleBinding(ctx, controllerContext, role, rolebinding, ns, func() ([]*workapiv1.ManifestWork, error) {
+		return m.manifestWorkLister.ManifestWorks(ns.Name).List(labels.Everything())
+	})
+
+	if err != nil {
+		klog.Errorf("Reconcile role/rolebinding %s fails with err: %v", key, err)
+	}
+	return err
+}
+
+func (m *FinalizeController) syncRoleAndRoleBinding(ctx context.Context, controllerContext factory.SyncContext,
+	role *rbacv1.Role, rolebinding *rbacv1.RoleBinding, ns *corev1.Namespace, listWorksFunc listManifestWorksFunc) error {
+	klog.V(4).Infof("Sync role/rolebinding %s/%s%s", ns.Name, ns.Name, keySuffix)
+	// no work if neither role nor rolebinding exists
+	if role == nil && rolebinding == nil {
+		return nil
+	}
+
+	// no work if no finalizer found on either role or rolebinding
+	if !hasFinalizer(role, rolebinding) {
+		return nil
+	}
+
+	if ns.DeletionTimestamp == nil || ns.DeletionTimestamp.IsZero() {
+		// ignore role if it is not terminating
+		if role != nil && (role.DeletionTimestamp == nil || role.DeletionTimestamp.IsZero()) {
+			role = nil
+		}
+
+		// ignore rolebinding if it is not terminating
+		if rolebinding != nil && (rolebinding.DeletionTimestamp == nil || rolebinding.DeletionTimestamp.IsZero()) {
+			rolebinding = nil
+		}
+	} else {
+		// clean up manifest works in the terminating namespace
+		done, err := m.cleanUpManifestWorks(listWorksFunc)
+		if err != nil {
+			return err
+		}
+
+		// requeue the key if there exists any manifest work in cluster namespace
+		if !done {
+			key := controllerContext.QueueKey()
+			controllerContext.Queue().AddAfter(key, RequeueDelay)
+			klog.V(4).Infof("Requeue role/rolebinding %q after %v", key, RequeueDelay)
+			return nil
+		}
+	}
+
+	// remove finalizer from role/rolebinding
+	errs := []error{}
+	if role != nil {
+		role = role.DeepCopy()
+		if changed := removeFinalizer(role, manifestWorkFinalizer); changed {
+			_, err := m.rbacClient.Roles(role.Namespace).Update(ctx, role, metav1.UpdateOptions{})
+			if err != nil {
+				errs = append(errs, err)
+			}
+		}
+	}
+
+	if rolebinding != nil {
+		rolebinding = rolebinding.DeepCopy()
+		if changed := removeFinalizer(rolebinding, manifestWorkFinalizer); changed {
+			_, err := m.rbacClient.RoleBindings(rolebinding.Namespace).Update(ctx, rolebinding, metav1.UpdateOptions{})
+			if err != nil {
+				errs = append(errs, err)
+			}
+		}
+	}
+
+	if len(errs) > 0 {
+		return utilerrors.NewAggregate(errs)
+	}
+	return nil
+}
+
+func (m *FinalizeController) cleanUpManifestWorks(listWorksFunc listManifestWorksFunc) (bool, error) {
```

Comment:
doc what this is doing for me.  I expected to simply count the manifests works in the namespace and if there was more than zero, don't remove the finalizer on the role or rolebinding.

Reply:
This func is removed along with the timeout feature.